lib/auth/userloginstate/generator.go

```diff
@@ -162,19 +163,23 @@ func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginst
 
 	if g.usageEvents != nil {
 		// Emit the usage event metadata.
-		if err := g.emitUsageEvent(ctx, user, uls); err != nil {
+		if err := g.emitUsageEvent(ctx, user, uls, inheritedGrants); err != nil {
 			g.log.Debug("Error emitting usage event during user login state generation, skipping")
 		}
 	}
 
 	return uls, nil
 }
 
-// addAccessListsToState will added the user's applicable access lists to the user login state.
-func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) error {
+// addAccessListsToState will add the user's applicable access lists to the user login state.
+func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) (accesslist.Grants, error) {
```

Comment:
This function is now a bit weird. It modifies login state and yet returns inherited grants for some reason. I understand you probably need them for the usage even emitted later. Can we add inherited grants to user state as well to keep logic consistent? Then you wouldn't need to modify any method signatures here either.

Reply:
Yeah, I modified its return for use in the event. Could do to just add them to the state.